Create debugger as a compiler helper

smfoote opened the pull request:
## What is included in this PR:
- The debugger helper now inserts a `debugger;` statement directly into the compiled template.
- Extensions can add instructions
- Instructions aren't really passes anymore - they happen after the passes are done (sort of like the code generator and post process).
  - This is necessary to allow extensions to add instructions. Before `buildInstructions` was one of the extensions, and allow extensions to extend extensions would have been a little too wild.
- The compiler now has a `ready` method that should be called before the compiler is used. The `ready` method is called after all of the extensions have been added by the application code.

smfoote committed:
Move debugger helper to the compiler

smfoote committed:
Allow instructions to be added by extensions

An extension can now modify a node in the AST, then create an instruction
to handle the modified node.

smfoote committed:
Make code generator functions extensible

prashn64 commented on test/sandbox/sandbox.js:
```diff
@@ -11,6 +12,12 @@ var compiledContainer = document.querySelector('#output .compiled');
+
+compiler.useExtension(tdDebugger);
+
+compiler.ready();
```
Can we change this API so that useExtension isn't even a thing.  This way when you call compiler.useExtensions, it will call ready on it's own right after.

smfoote committed:
Don't externalize the compiler.ready method

This method can be called as many times as needed, so we can call it
after each extension is added, without any negative effects. I am
removing it from the public API, therefore, per Prash's suggestion.
